src/generic/AbsMSeries.jl

```diff
@@ -0,0 +1,774 @@
+
+Return a vector of precision caps, one for each variable in the ring.
+Arithmetic operations will be performed to precisions not exceeding these
+values.
+"""
+max_precision(R::AbsMSeriesRing) = R.prec_max
+
+@doc Markdown.doc"""
+    valuation(a::AbsMSeries)
+
+Return the valuation of $a$ as a vector of integers, one for each variable.
+"""
+function valuation(a::AbsMSeries)
+    p = poly(a)
+    prec = precision(a)
+    val = prec
+    for v in exponent_vectors(p)
+        if iszero(v)
+            return v
+        end
+        v = exponents_clamp_zero_to_prec(v, prec)
+        val = min.(v, val)
+    end
+    return val
+end
+
+@doc Markdown.doc"""
+    coeff(a::AbsMSeries, n::Int)
+
+Return the coefficient of the $n$-th nonzero term of the series (or zero if
+there are fewer than $n$ nonzero terms). Terms are numbered from the least
+significant term, i.e. the first term displayed when the series is printed.
+"""
+function coeff(a::AbsMSeries, n::Int)
+    return coeff(poly(a), length(a) - n + 1)
+end
+
+iszero(a::AbsMSeries) = length(poly(a)) == 0
+
+function isone(a::AbsMSeries)
+    return isone(poly(a)) || iszero(precision(a))
+end
+
+zero(R::AbsMSeriesRing) = R(0)
+
+one(R::AbsMSeriesRing) = R(1)
+
+@doc Markdown.doc"""
+    isunit(a::AbsMSeries)
+
+Return `true` if the series is a unit in its series ring, i.e. if its constant
+term is a unit in the base ring.
+"""
+isunit(a::AbsMSeries) = isunit(constant_coefficient(poly(a)))
+
+@doc Markdown.doc"""
+    gen(R::AbsMSeriesRing, i::Int)
+
+Return the $i$-th generator (variable) of the series ring $R$. Numbering starts
+from $1$ for the most significant variable.
+"""
+function gen(R::AbsMSeriesRing, i::Int)
+    S = R.poly_ring
+    prec = [R.prec_max[ind] for ind in 1:nvars(R)]
+    x = R.prec_max[i] > 1 ? gen(S, i) : S()
+    return R(x, prec)
+end
+
+@doc Markdown.doc"""
+    gens(R::AbsMSeriesRing)
+
+Return a vector of the generators (variables) of the series ring $R$, starting
+with the most significant.
+"""
+gens(R::AbsMSeriesRing) = [gen(R, i) for i in 1:nvars(R)]
+
+
+@doc Markdown.doc"""
+    isgen(a::AbsMSeries)
+
+Return true if the series $a$ is a generator of its parent series ring.
+"""
+function isgen(a::AbsMSeries)
+    R = parent(a)
+    p = poly(a)
+    v = vars(p)
+    return length(v) == 1 && length(p) == 1 &&
+                          isone(lc(p)) && sum(first(exponent_vectors(p))) == 1
+end
+
+@doc Markdown.doc"""
+    vars(R::AbstractAlgebra.MSeriesRing)
+
+Return a vector of symbols, one for each of the variables of the series ring
+$R$.
+"""
+vars(R::AbstractAlgebra.MSeriesRing) = R.sym
+
+parent(a::AbstractAlgebra.MSeriesElem) = a.parent
+
+function base_ring(R::AbstractAlgebra.MSeriesRing{T}) where T <: RingElement
+    return R.base_ring::parent_type(T)
+end
+
+base_ring(a::AbstractAlgebra.MSeriesElem) = base_ring(parent(a))
+
+function deepcopy_internal(a::AbsMSeries, dict::IdDict)
+    return parent(a)(deepcopy(poly(a)), precision(a))
```

Comment:
Maybe pass recursively `dict`, in a recursive call to `deepcopy_internal` instead of `deepcopy`?

Reply:
Done.